### DISCUSSION TITLE
Why switch to LibreSSL (was: BoringSSL)?

### DISCUSSION THREAD

**Discussion by jay** [curl/curl-for-win#44] [Q&A] [answered]
Why switch the default SSL lib to BoringSSL, I would think we'd want to keep OpenSSL the default because like 90% of curl users or something (I forget the exact stats) use it. Is it really that similar to OpenSSL? The path vulnerability you cite was fixed long ago wasn't it?

**Comment by vszakats** [comment, marked as answer]
No, the vulnerability was never fixed. It was merely mitigated by moving the settings under `C:\Program Files\` (from `/usr/local/` and similar paths that were used before by default), which remains vulnerable on all localized Windows versions that use a [different folder name](https://en.wikipedia.org/wiki/Program_Files#Localization). I could mitigate it more by moving the settings under `C:\Windows\System32\`, but even this is not safe as the OS might be installed in a different directory. Even for this I must patch OpenSSL's `Configure`, because there is a cross-build bug which prevents overriding the settings directory, saying that the directory isn't absolute when it is. For this bug I provided a patch, which they refused to merge (asking for personal data to sign another CLA, because someone deemed the _two-liner_ fix, that I finally contributed myself, as "non-trivial"). They also refuse to fix the vulnerability itself because "they do not provide Windows binaries".

The correct fix is to either use a runtime-detected Windows OS location as a base directory for its settings, or even better would be to allow disabling the loading of settings from disk altogether (as BoringSSL does for example). LibreSSL has the `C:\Windows\` mitigation in place, out of the box. [Aside: `libssh` suffers from similar issues on Windows]

Other reasons:
- curl-for-win already doesn't use OpenSSL, but a fork of it, see next:
- No HTTP/3 support. So one must rely on a fork, such as `quictls`. Which in turn isn't GPG signed (or even checksummed, or formally released). UPDATE: fixed by v3.3.1, but with significantly lower performance and higher memory requirement than `quictls` (and other forks.)
- Large binaries. (BoringSSL is significantly smaller. Exact figures are in the commits [\[1\]](https://github.com/curl/curl-for-win/commit/8a3a331334f7322cc21670d364db4b3f99ebbb8f) [\[2\]](https://github.com/curl/curl-for-win/commit/c90c3c1ea5b9ddeed8b0f87afe2aff6dee0adb35) from the time when I was dealing with this)
- Regressions (from 2022: 3.0.3 requiring rollback due case-sensitive comparison bug, 3.0.6 withdrawn for a CRITICAL vulnerability fix.)
- Rejected Windows vulnerability fix with loading configuration from hardcoded paths. (see above). UPDATE: around 2024/2025 OpenSSL released a mitigation making it possible to override the hard-coded config path, via HKLM registry, per runner machine, by an administrator, "at installation". OpenSSL remains vulnerable by default, with my expressed concerns rejected, also refusing to recognize it as a bug: https://github.com/openssl/openssl/issues/24528
- Unfixed cross-build bug, that prevents mitigation of the above. (see above)
- QUIC API support not planned. A different solution is planned and even that is years away. UPDATE: solved by v3.5.0, released in April 2025, by providing a QUICTLS-like API.
- Speaking of the planned "different solution": It seems risky to rely a newly-designed, supported-by-no-app, networked QUIC component added to an already large and fragile OpenSSL codebase. That was (and still is) the general sentiment after these plans were revealed. UPDATE: turned to be a valid concern.
- ~Missing Windows ARM64 ASM support~. (to be fair this also applies to the alternatives). Fixed with OpenSSL v3.1.0.
- Uses proprietary build system.
- Requires a patch for reproducible builds. (otherwise certain build settings/properties are included in the final binaries)
- Requires build tricks to disable dynamic loading of code.
- Windows support feels like an afterthought.
- Locking/performance issues with 3.x:
  - https://news.ycombinator.com/item?id=33831967
  - https://github.com/openssl/openssl/issues/20286
  - https://github.com/haproxy/haproxy/issues/2454#issuecomment-2353243899
  - https://www.haproxy.com/blog/state-of-ssl-stacks (2025)
  - https://www.memorysafety.org/blog/rustls-server-perf/ (2025) and https://rustls.dev/perf/2024-11-28-threading/
  - https://www.memorysafety.org/blog/rustls-performance-outperforms/
- Uses a kernel header on Linux: `linux/mman.h`. It makes cross-building much more complicated than expected, and impossible to fix in certain cases. Workaround is to build with `no-secure-memory`, which does not sound good.

Since the end of 2022, both LibreSSL 3.7.0 and BoringSSL are fully on par with OpenSSL (or rather with the `quictls` fork!), with the exception of `TLS-SRP`, which is missing from both. SRP is old and [not recommended](https://blog.cryptographyengineering.com/should-you-use-srp/) nowadays. Also since a couple of weeks, curl [started using a function](https://github.com/curl/curl/pull/10151#discussion_r1058475335) which is missing from LibreSSL. This is going to affect 7.88.0, though it's unclear in what ways [UPDATE: it's an internal detail, with no user-facing effect]. LibreSSL 3.8.1 [will drop](https://github.com/libressl/portable/blob/aab9622a3fbfad3236d761e92fffd5bc17f0d698/ChangeLog#L68-L72) loading of engines, conf modules and disables DSO altogether.

I shall add, that the cited vulnerability shouldn't affect curl itself, because we build it with `-DCURL_DISABLE_OPENSSL_AUTO_LOAD_CONFIG`. I haven't tested what else may still be loaded from the settings directories, plus we're distributing OpenSSL static libs, which remain affected by the vulnerability.

Overall, I think most curl-for-win users are using `curl.exe`, for them, the choice of TLS-backend is transparent for the most part. It may matter for static lib users, but I'd guesstimate their numbers not too high. And by that I mean _very low_.

**Reply by vszakats** [reply]
Another opinion on the state of OpenSSL:
https://github.com/haproxy/haproxy/issues/680#issuecomment-1433118828

**Reply by jay** [reply]
Interesting read. I hope things improve.

**Comment by vszakats** [comment]
Since LibreSSL (with v3.7.x) has caught up in features with BoringSSL, and BoringSSL takes quite a bit of effort to adjust for our purposes, and also doesn't seem that its pthread requirement is going away anytime soon (hitting a toolchain bug, which also seems to stay with us), I plan to switch to LibreSSL instead.

This will result in significantly smaller binaries, faster builds, hack- and patch-free build process and a proven development path with versioned releases. Also worth a note that the latest OpenSSL release 3.1.1 from 2 months ago, still did not get a matching release from the quictls fork. Instead it got a second set of patches for performance. Meanwhile mainline OpenSSL turned to focus almost exclusively on its own in-house QUIC implementation.

**Comment by vszakats** [comment]
@arm64-v9a _That_ QUIC is OpenSSL's home-grown implementation. It isn't compatible with curl or any software out there and it isn't released yet. A missing quictls-compatble API is just one in the list of concerns with OpenSSL.

There is no perfect choice out there, including the mainline. E.g. OpenSSL itself suffers from a significant performance drop since 3.x, due to locking.

**Reply by vszakats** [reply]
A quictls QUIC API is required for HTTP/3 in curl. OpenSSL decided that it will not provide one.

**Reply by vszakats** [reply]
After playing with LibreSSL for a while today, found some caveats:

LibreSSL offers ASM support for x64 only (which is OK). ASM is enabled by default in autotools builds, but due to bugs, it's never enabled in CMake MinGW builds (MSVC seems OK, but haven't tested). The next problem is that 3.8.0 builds are consistently broken with ASM enabled (they do not even start, as if the exe was corrupted). 3.7.3 works fine.

**Reply by vszakats** [reply]
Tried it, but llvm [doesn't support](https://reviews.llvm.org/D99565) this option, so failed at `./configure`: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/47725397#L4569

**Reply by vszakats** [reply]
PR for the LibreSSL CMake/MinGW/ASM bugs: https://github.com/libressl/portable/pull/894
Related Issue also for the 3.8.0 crashes: https://github.com/libressl/portable/issues/893

**Reply by vszakats** [reply]
Raw crypto performance is important, but not the most important property of a TLS-backend in our primary use-case of transferring data over networks either manually or via shell scripts.

OpenSSL is unbeatable in raw crypto perf as they've been focusing on this heavily. It's also one reason why the library grew so huge in size. E.g. `libcrypto-lib-aes-gcm-avx512.obj` in 3.1.2, a single function, at 729KB, is larger than libcurl itself in a perfectly usable configuration. This to me seems overkill for our purposes.

It'd be interesting to see performance comparison on typical hardware/network with typical workloads and typical, secure servers using curl built against various backends. If performance turns out to be remarkably worse, it's probably better to raise these issues directly with LibreSSL.

curl-for-win also offers the Schannel backend by default. It's an option for cases the default backend performs worse than desired. Custom builds also remain possible with all popular backends.

**Comment by vszakats** [comment]
Status updates: 
- LibreSSL 3.8.2 stable [is out](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.8.2-relnotes.txt). It includes some necessary build patches. Also fixes the Windows x64 crash.
https://github.com/libressl/portable/pull/894
https://github.com/libressl/portable/pull/909
- Still looking forward for some last-minute LibreSSL build patches coming up in the next release. No showstoppers.
https://github.com/libressl/portable/pull/940
https://github.com/libressl/portable/pull/938
https://github.com/libressl/portable/pull/937
https://github.com/libressl/portable/pull/936
https://github.com/libressl/portable/pull/932
https://github.com/libressl/portable/pull/931
https://github.com/libressl/portable/pull/930
https://github.com/libressl/portable/pull/927
- LibreSSL CMake builds are ready for use and have been tested extensively.
- There is also a short list of LibreSSL [caveats](https://github.com/curl/curl-for-win/blob/d14b657c30d57ad0b87f87c26e5a8ed3cca42755/libressl-pkg.sh#L6-L25), e.g. [missing support](https://github.com/libressl/portable/issues/838) for `SSL_set0_wbio()`. But also, no showstoppers here.
- Pushed a series of curl patches to enable certain OpenSSL codepaths also for LibreSSL. Some of these were out with 8.4.0, some more coming with 8.5.0.
https://github.com/curl/curl/pull/12385
https://github.com/curl/curl/pull/12274
https://github.com/curl/curl/pull/11616
https://github.com/curl/curl/pull/11615
https://github.com/curl/curl/pull/11611
https://github.com/curl/curl/pull/11555
- and to libssh2:
https://github.com/libssh2/libssh2/pull/1146
- and to ngtcp2:
https://github.com/ngtcp2/ngtcp2/issues/935
https://github.com/ngtcp2/ngtcp2/pull/936
- OpenSSL 3.2.0 was [released today](https://github.com/openssl/openssl/blob/openssl-3.2.0/CHANGES.md#openssl-32).
- One long-awaited 3.2.0 feature for Windows is support
  for the system certificate store (not yet enabled by default
  at runtime). I believe curl supports that also for LibreSSL
  via `--ca-native` and `--proxy-ca-native` and via
  `CURLSSLOPT_NATIVE_CA` in libcurl.
- It seems that quictls stops rebasing on top of 3.2.0.
  Meaning stuck with 3.1.x bugfix releases.
  https://github.com/quictls/openssl/issues/138
- Path for future quictls updates is unclear. Current best option
  may be coming with OpenSSL 3.3, pending questions:
  https://github.com/quictls/openssl/discussions/124

I don't expect much to change short/mid-term with the quictls
(or OpenSSL QUIC) situation. So, unless I missed something here
or a last-minute showstopper comes up, it means I'm making the
switch to LibreSSL on the next curl release, 8.5.0. Preceded by
a final revision of 8.4.0 built with LibreSSL, to be able to sort out
if a possible regression is due to the TLS backend change or due
to the curl version update.
